### PR TITLE
Task API integration test script

### PR DESCRIPTION
**Pull request by Krigpl**
Just the bug in the Task API startup is an issue here (those time.sleep calls).

### CODE REVIEW

File: golem/task/requestedtaskmanager.py [hunk 1]
```diff
@@ -360,6 +368,8 @@ def get_subtasks_outputs_dir(self, task_id: TaskId) -> Path:
             logger.info(
                 'app_client created for app_id=%r, clients=%r',
                 app_id, self._app_clients[app_id])
+            import time
+            time.sleep(2)
```
Comment (Wiezzel) [inline golem/task/requestedtaskmanager.py]: 😿

File: golem/task/task_api/__init__.py [hunk 2]
```diff
@@ -70,3 +70,4 @@ def running(self) -> bool:
         assert self._runtime is not None
         loop = asyncio.get_event_loop()
         await self._runtime.wait_until_stopped().asFuture(loop)
+        await self._runtime.clean_up().asFuture(loop)
```
Comment (Wiezzel) [inline golem/task/task_api/__init__.py]: How about this?
```python3
try:
    await self._runtime.wait_until_stopped().asFuture(loop)
finally:
    await self._runtime.clean_up().asFuture(loop)
```

File: golem/task/taskcomputer.py [hunk 3]
```diff
@@ -320,6 +320,8 @@ def compute(self) -> defer.Deferred:
         )
 
         self._app_client = await ProviderAppClient.create(task_api_service)
+        import time
+        time.sleep(2)
```
Comment (Wiezzel) [inline golem/task/taskcomputer.py]: 😿

File: scripts/task_api_tests/basic_integration.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+from golem.task.task_api import docker
+from golem.envs.auto_setup import auto_setup
+from golem.envs.docker import cpu, whitelist
+
+from twisted.internet.task import react
+from twisted.internet.defer import ensureDeferred, Deferred
+
+import logging
+logging.basicConfig(level=logging.INFO)
+# logging.basicConfig(level=logging.DEBUG)
```
Comment (Wiezzel) [inline scripts/task_api_tests/basic_integration.py]: Plz, remove commented line. Also the import order looks weird.

File: scripts/task_api_tests/basic_integration.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    )
+    env_manager.set_enabled(environment, True)
+
+    env_prerequisites = json.loads(env_prerequisites_json)
+
+    rtm_work_dir = work_dir / 'rtm'
+    rtm_work_dir.mkdir()
+    rtm = requestedtaskmanager.RequestedTaskManager(
+        env_manager,
+        b'',
```
Comment (Wiezzel) [inline scripts/task_api_tests/basic_integration.py]: Better pass parameters by keywords here because this empty string looks obscure.

File: golem/task/requestedtaskmanager.py (hunk 1 above)
Comment (Krigpl) [reply golem/task/requestedtaskmanager.py]: Dunno what to do about it right now, tbh

Comment (Wiezzel) [reply golem/task/requestedtaskmanager.py]: Me neither, but this makes me sad.

File: scripts/task_api_tests/basic_integration.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+from golem.envs.docker import cpu, whitelist
+
+from twisted.internet.task import react
+from twisted.internet.defer import ensureDeferred, Deferred
+
+logging.basicConfig(level=logging.INFO)
+
+
+def as_deferred(f):
+    return Deferred.fromFuture(asyncio.ensure_future(f))
```
Comment (Wiezzel) [inline scripts/task_api_tests/basic_integration.py]: Plz, use `golem.core.deferred_from_future`
https://github.com/golemfactory/golem/blob/239d47d5953832d287302b2aad1bb210df29fa69/golem/core/deferred.py#L67-L68

Comment (Wiezzel) [reply scripts/task_api_tests/basic_integration.py]: Actually, you can modify `golem.core.deferred_from_future` to accept any `Awaitable` object by putting `ansycio.ensure_future()` there.

File: golem/task/requestedtaskmanager.py [hunk 7]
```diff
@@ -362,6 +362,8 @@ def get_subtask_outputs_dir(self, task_id: TaskId) -> Path:
...
```
Comment (maaktweluit) [inline golem/task/requestedtaskmanager.py]: Also makes me sad, maybe we can at least use asyncio?

```suggestion
            await asyncio.sleep(2)
```

File: golem/task/taskcomputer.py (hunk 3 above)
Comment (maaktweluit) [reply golem/task/taskcomputer.py]: same as the other sleep:
:( and async?

File: scripts/task_api_tests/basic_integration.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        resources=resources,
+        max_subtasks=max_subtasks,
+        max_price_per_hour=1,
+        concent_enabled=False,
+    )
+    with open(task_params_path, 'r') as f:
+        task_params = json.load(f)
+    task_id = rtm.create_task(golem_params, task_params)
+    print('Task created', task_id)
+    await deferred_from_future(rtm.init_task(task_id))
```
Comment (maaktweluit) [inline scripts/task_api_tests/basic_integration.py:91]: why do you need a deferred here? i think `await rtm.init_task(..)` would work to

File: scripts/task_api_tests/basic_integration.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+
+async def _task(
+        environment,
+        env_prerequisites,
+        task_params_path,
+        resource,
+        max_subtasks,
+) -> None:
+    work_dir = Path(tempfile.mkdtemp())
```
Comment (maaktweluit) [inline scripts/task_api_tests/basic_integration.py]: maybe we can use `tmp_path` fixture from pytest here too?

File: scripts/task_api_tests/basic_integration.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    print('work_dir', work_dir)
+    db = database.Database(
```
Comment (maaktweluit) [inline scripts/task_api_tests/basic_integration.py:156]: we can use `AsyncDatabaseFixture` for this too https://github.com/golemfactory/golem/pull/4648/files#diff-70d60ac05925af68aa36648b86338e99R181

**Review by maaktweluit — CHANGES_REQUESTED** [review]
LGTM!

Only the deferred q is important, the others are minor.

I know this is not using `pytest` now, but think it would be nice to share common components between tests.

**Comment by Krigpl** [comment]
@maaktweluit The reason I didn't use pytest is that it needs to take command lint arguments to run. Does pytest support that?

File: golem/task/requestedtaskmanager.py (hunk 7 above)
Comment (Krigpl) [reply golem/task/requestedtaskmanager.py]: Maybe let's just move it to the Task API repo since it's an issue there.

Comment (maaktweluit) [reply golem/task/requestedtaskmanager.py]: 👍

**Comment by maaktweluit** [comment]
@Krigpl yes 
you need to add them to the options in `conftest.py`, like https://github.com/golemfactory/golem/blob/develop/scripts/node_integration_tests/conftest.py#L39

File: golem/task/requestedtaskmanager.py (hunk 7 above)
Comment (mfranciszkiewicz) [reply golem/task/requestedtaskmanager.py]: Server health check may help here https://github.com/golemfactory/task-api/pull/36

File: scripts/task_api_tests/basic_integration.py (hunk 8 above)
Comment (Wiezzel) [reply scripts/task_api_tests/basic_integration.py:91]: Yea, this is a very good point. Why do you even need deferreds when the method is `async`?

Comment (Krigpl) [reply scripts/task_api_tests/basic_integration.py:91]: This is some magic related to incorporating twisted and asyncio together.
I followed this article https://meejah.ca/blog/python3-twisted-and-asyncio
It doesn't work without converting to deferred.
tl;dr; of the article is that you need to be aware of what context are you really in when inside an async function - twisted vs asyncio.

**Comment by Krigpl** [comment]
@maaktweluit I find it a little awkward to introduce global flag available for 1k+ tests for the sake of a single test. Can still do it if you think it's worth it.

**Comment by Krigpl** [comment]
Also this isn't meant to be run together with the rest of the test suite. It just doesn't make much sense to me to use pytest here.

**Comment by Wiezzel** [comment]
> Also this isn't meant to be run together with the rest of the test suite. It just doesn't make much sense to me to use pytest here.

This is a fair point. It was meant to be a tool for developing apps, not for testing the Golem codebase.

File: scripts/task_api_tests/basic_integration.py (hunk 8 above)
Comment (maaktweluit) [reply scripts/task_api_tests/basic_integration.py:91]: ahh, clear, had that fight a few times too the last weeks

maybe we can remove twisted from this test, would be nice to have it native asyncio

**Comment by maaktweluit** [comment]
> @maaktweluit I find it a little awkward to introduce global flag available for 1k+ tests for the sake of a single test. Can still do it if you think it's worth it.

@Krigpl agree with that, but the `conftest.py` file can also be applied locally ( like the linked sample for `node_integration_tests` )

> Also this isn't meant to be run together with the rest of the test suite. It just doesn't make much sense to me to use pytest here.

Agree with this too, but its another test inside the golem-core repo, and most other tests are compatible to run with pytest ( unit, node_e2e, concent_e2e )

I do not mind if we use pytest here or not, the main point of my comment was to be able to re-use test utilities, since `tmpdir` and `DatabaseFixture` are implemented again in this file.

File: scripts/task_api_tests/basic_integration.py (hunk 8 above)
Comment (Krigpl) [reply scripts/task_api_tests/basic_integration.py:91]: I doubt we can remove twisted here completely considering the dependent components still use it, i.e. twisted reactor still has to be installed for some golem components to work properly.

**Comment by Krigpl** [comment]
@maaktweluit Right, but another thing is that I am considering making `work_dir` a flag, since it can't be tmpdir in Windows (because of docker mounts) and whether to remove it automatically also configurable by a flag. Thus not reusable from pytest anymore.

File: scripts/task_api_tests/basic_integration.py (hunk 8 above)
Comment (maaktweluit) [reply scripts/task_api_tests/basic_integration.py:91]: ok, that is a bummer, lets leave it then for now :)

**Review by maaktweluit — APPROVED** [review]
lets leave it without pytest and with twisted for now

if we want to move there later the `tmpdir` is also a flag there

File: golem/task/requestedtaskmanager.py (hunk 7 above)
Comment (maaktweluit) [reply golem/task/requestedtaskmanager.py]: that will solve it :)
